Adds support for downloading biological assemblies to Bio.PDB.PDBList

sean-workman opened the pull request:
<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->

Closes #1875 

This PR adds the ability to request assemblies for download along with regular entries. It requests the assemblies from the newly available directory (as of May 2022) that provides assemblies in mmCIF format for all PDB entries. I am working on putting together a couple of unit tests now. This is my first contribution and I am grateful to @JoaoRodrigues for providing some help and guidance through the process.

peterjc commented:
The pre-commit.ci checks say you need to run black on some of the Python files - see https://github.com/biopython/biopython/blob/master/CONTRIBUTING.rst for how to do these checks on your own machine.

sean-workman commented:
Thanks @peterjc, that was a silly mistake on my end. I was getting an ImportError from black but sorted it out with an update of pre-commit itself. It appears that the PR is failing on CircleCI now, but it looks like the failure isn't anything related to the PDBList module.

JoaoRodrigues commented on .pre-commit-config.yaml:
```diff
@@ -5,7 +5,7 @@
 #
 repos:
```
We should remove these changes from the PR before merging! Not sure if they are already in the main repo or if they were the product of running pre-commit autoupdate or so.

JulianMaurin replied:
👍, I would favour limiting the scope of the PR to `PDBList`.
It would be a good idea to update the pre-commit dependencies version, but it can be done in a dedicated PR.

JulianMaurin replied:
You can easily revert this commit (40b0c48d392db0dd3da3ac6c44ae350191742008).
To do so: `git revert 40b0c48d392db0dd3da3ac6c44ae350191742008`.
Notice, that the commit contains also a `print` update, that could be re-apply individually.
_(If you want to challenge your Git skills choosing the best pratice, have a look at the [interactive rebase](https://jwiegley.github.io/git-from-the-bottom-up/1-Repository/8-interactive-rebasing.html))_

JoaoRodrigues commented on Bio/PDB/PDBList.py:
```diff
@@ -364,12 +365,27 @@ def update_pdb(self, file_format=None):
+
+            except Exception as err:
+                print(f"error {pdb_code}: {err}\n")
                 # you can insert here some more log notes that
```
Maybe remove these comments?

JoaoRodrigues commented on NEWS.rst:
```diff
@@ -53,6 +53,7 @@ possible, especially the following contributors:
 - Peter Cock
```
I'd add a small blurb to the top of the file mentioning this change, something along the lines of: "PDBList now allows for the download of biological assemblies, in PDB and mmCIF format."

JoaoRodrigues commented:
By the way, the failing test in CircleCI is unrelated to this PR so for me, we're good to merge.

JulianMaurin reviewed (commented):
Hello @sean-workman 👋 
Good job ! 
Especially the integration of the `ftplib`. I been working on the introduction of a new module `PDBServer` (see: #4012), I feel like it would be nice to centralize the FTP capabilities there at a point. 

Despite all these good things, I think it would be necessary to add some test cases.
I can help if needed, don't hesitate.

sean-workman commented:
Hi @JulianMaurin, thank you very much, it was really helpful to have Joao giving guidance. I have been preparing for a job interview so I've been a bit distracted, but I was going to have a go at some writing some unit tests today.

I will see how it goes but I certainly won't hesitate to ask for help if I need it. I'm sure whatever I come up with will likely need a bit of tidying up.

sean-workman commented:
Hey @JulianMaurin, it turns out I might need a hand sooner than expected. For some reason I am getting an AttributeError saying that the new methods (get_all_assemblies(), etc.) don't exist. 

It almost seems like unittest is trying to look at the main branch version of PDBList, but I can mouse over the method calls and see the docstrings for the new methods, so not entirely sure what is going on. The simple test below is what I've tried adding to start:

```
    def test_get_all_assemblies(self):
        """Tests the Bio.PDB.PDBList.get_all_assemblies method."""
        # obsolete_pdb declared to prevent from creating the "obsolete" directory
        pdblist = PDBList(obsolete_pdb="unimportant")
        entries = pdblist.get_all_assemblies()
        # As number of obsolete entries constantly grow, test checks if a certain number
        # was exceeded
        self.assertGreater(len(entries), 100000)
```

And what I get as an output is:

```
ERROR: test_get_all_assemblies (__main__.TestPBDListGetList)
Tests the Bio.PDB.PDBList.get_all_assemblies method.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sean/python/biopython/Tests/test_PDBList.py", line 56, in test_get_all_assemblies
    entries = pdblist.get_all_assemblies()
AttributeError: 'PDBList' object has no attribute 'get_all_assemblies'

----------------------------------------------------------------------
```

This is my first time using the unittest library and I've tried looking around for people with similar issues but I've not come up with a solution yet.

JulianMaurin commented:
Hey @sean-workman 👋 

The test is working fine on my side, based on your last commit.
Lest's try by adding only piece of code:

**Tests/test_PDBList.py (right after `TestPDBListGetStructure`)**

```python
class TestPDBListAssembly(unittest.TestCase):
    def test_get_all_assemblies(self):
        """Tests the Bio.PDB.PDBList.get_all_assemblies method."""
        # obsolete_pdb declared to prevent from creating the "obsolete" directory
        pdblist = PDBList(obsolete_pdb="unimportant")
        entries = pdblist.get_all_assemblies()
        # As number of obsolete entries constantly grow, test checks if a certain number
        # was exceeded
        self.assertGreater(len(entries), 250000)
```

sean-workman commented:
I've now added in tests for the new methods in PDBList. Please let me know if there are any that you think I may have missed. I used the same PDB codes as in the previous tests for consistency and used the existing tests as a guide, but right after pushing these changes I realized including a "large" mmCif assembly file doesn't really achieve much since the whole point before was the test the "bundle" option.

In the course of making and running these tests I noticed a small error in the formatting of file names for assemblies in .pdb format, so I went back and fixed that in the PDBList module itself.

JulianMaurin commented on Tests/test_PDBList.py:
```diff
@@ -134,13 +143,89 @@ def test_retrieve_pdb_file_mmtf(self):
+                structure, assembly_num, pdir=pdir, file_format=file_format
+            )
+            self.assertTrue(os.path.isfile(path))
+            os.remove(path)
```
You don't need to remove the file explicitly, the temp directory will get removed exiting the `make_temp_directory` decorator

sean-workman replied:
Great! I've removed that line in my TestPDBListGetAssembly tests as well as the existing TestPDBListGetStructure tests.

JulianMaurin commented on Tests/test_PDBList.py:
```diff
@@ -134,13 +143,89 @@ def test_retrieve_pdb_file_mmtf(self):
+
+    def check(self, structure, assembly_num, filename, file_format, pdir=None):
+        with self.make_temp_directory(os.getcwd()) as tmp:
+            pdblist = PDBList(pdb=tmp, obsolete_pdb=os.path.join(tmp, "obsolete"))
```
```suggestion
            pdblist = PDBList(pdb=tmp)
```

The obsolete directory is build the same way in the `PDBList` constructor ([see](https://github.com/biopython/biopython/blob/c067ac0fcbd0a9ee5d19f3156f8ae299d4da11f0/Bio/PDB/PDBList.py#L108)).

JulianMaurin reviewed (approved):
Well done! I look better with the tests :clap:
Everything is not perfect but let's make it evolve step by step.

Thanks for the contribution

JulianMaurin commented:
ps: I think rebasing the master will fix the failing test (`git rebase master).

sean-workman commented:
Well... I tried rebasing and I'm going to go ahead and assume I did it wrong since it didn't fix the CircleCI test. I've never done a rebase before so not surprising. But the other changes have been incorporated!

JulianMaurin commented:
> Well... I tried rebasing and I'm going to go ahead and assume I did it wrong since it didn't fix the CircleCI test. I've never done a rebase before so not surprising. But the other changes have been incorporated!

Ah yes, sorry my message was incomplete.
You need to sync your master with this upstream one (eg. [here](https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a)).

_ps: It is recommended always to use `--force-with-lease` over `--force` using `git push`_

sean-workman commented:
Well, rebasing took me longer than I would have liked to figure out, but hey, learning! Finally looks like everything is passing. :)

JoaoRodrigues reviewed (approved):
LGTM! Very nice @sean-workman, thank you! Thanks for the comments and help @JulianMaurin ! If the tests pass, we're good to merge.
